File: Java/113-PathSumII.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    public List<List<Integer>> pathSum(TreeNode root, int sum) {
        List<List<Integer>> res =  new ArrayList<>();
        List<Integer> currPath = new ArrayList<>();
        findPathSum(root, res, currPath, sum);
        return res;
    }
    private void findPathSum(TreeNode root, List<List<Integer>> res, List<Integer> currPath, int sum){
        if (root == null){
            return;
        }
        currPath.add(root.val);
        if (root.left == null && root.right == null && root.val == sum){
            res.add(new ArrayList(currPath));
        }else{
            findPathSum(root.left, res, currPath, sum-root.val);
            findPathSum(root.right, res, currPath, sum-root.val);
        }
        currPath.remove(currPath.size()-1);
    }
}

File: cpp/113-PathSumII.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    vector<vector<int>> pathSum(TreeNode* root, int sum) {
        vector<vector<int>> res;
        if (!root){
            return res;
        }
        vector<int> curr;
        dfs(root, sum, res, curr);
        return res;
    }

    void dfs(TreeNode* root, int sum, vector<vector<int>>& res, vector<int>& curr){
        if (!root){
            return;
        }
        curr.push_back(root->val);
        if (root->val == sum && !root->left && !root->right){
            res.push_back(curr);
        }
        dfs(root->left, sum - root->val, res, curr);
        dfs(root->right, sum - root->val, res, curr);
        curr.pop_back();
    }
};
